Add ability to register prehooks to grad_fn

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83331
* __->__ #83226
* #83225

This simply replicates the implementation of PyFunctionPostHooks

Fixes https://github.com/pytorch/pytorch/issues/83120

soulitzer committed:
Add ability to register prehooks to grad_fn

[ghstack-poisoned]

soulitzer commented on test/test_autograd.py:
```diff
@@ -766,6 +766,37 @@ def bw_hook_modify(grad):
+        b.sum().backward()
+
+        self.assertEqual(post_counter[0], 2)
+
```
More tests:
- remove hooks
- remove hook during backward
- register hooks on tensor as well

albanD reviewed (commented):
Approach sounds good. I think we could avoid quite a bit of code duplication though.

soulitzer committed:
Update on "Add ability to register prehooks to grad_fn"


This simply replicates the implementation of PyFunctionPostHooks

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/python_hook.cpp:
```diff
@@ -75,6 +75,45 @@ auto PyFunctionTensorPreHook::operator()(const variable_list& values)
   return results;
 }
 
+PyFunctionPreHook::PyFunctionPreHook(PyObject* dict) : dict(dict) {
```
Good idea. I added a new PR in this stack to refactor the code to avoid modifying the two other APIs in this PR.

soulitzer committed:
Update on "Add ability to register prehooks to grad_fn"


This simply replicates the implementation of PyFunctionPostHooks

[ghstack-poisoned]

soulitzer committed:
Update on "Add ability to register prehooks to grad_fn"


This simply replicates the implementation of PyFunctionPostHooks

Fixes https://github.com/pytorch/pytorch/issues/83120

[ghstack-poisoned]

albanD reviewed (commented):
Looks good. Only the custom Function version is missing and this is good to go!

soulitzer committed:
Update on "Add ability to register prehooks to grad_fn"


This simply replicates the implementation of PyFunctionPostHooks

Fixes https://github.com/pytorch/pytorch/issues/83120

[ghstack-poisoned]

soulitzer commented:
Updated to also support custom Function!

albanD commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -929,6 +929,21 @@ PyObject* THPFunction_register_hook(PyObject* _self, PyObject* hook) {
+  auto cdata = self->cdata.lock();
+  TORCH_CHECK(
+      cdata,
+      "Attribute 'register_prehook' is invalid for this instance of _C._FunctionBase. "
```
Didn't we remove the python code for the old style custom Function? :p

soulitzer replied:
I guess.. but people can still do `Func().name()` lol

albanD reviewed (approved):
SGTM

soulitzer commented:
@pytorchbot merge -g
